### PR TITLE
Improve Tesseract module

### PR DESCRIPTION
**Pull request by bosd**
This is a rewrite of the tesseract implementation:
- Having a tesseract and tesseract 4 module was confusing. (basically it was both tesseract 4)
- Tesseract is now configured in auto mode
- Invoices are tested against the installed languages on the machine (alternative for #181)
- It will detect which engine is installed on your system.

- support for multipage PDF's as an input
- The pre-processor is only called on pdf input, image files are passed cleanly. #169 
- It will preserve the same layout as the input file.
Very usefull for when you lost the digital file, and only have a print on paper. You can scan it, and parse it now with the same pdftotext template. 

Fixes: 
Previous tesseract 4 implementation was broken.
Implementation of timeout error when conversion takes to long

### CODE REVIEW

**Comment by bosd** [comment]
Results, on a system without any additional languages or engine packs installed.
(Output can be improved when those are installed )

ooh ohh, end of my commute, will finish this post later...

**Comment by bosd** [comment]
## Sammy

**before**  sammymaystone.png as input


<details>


```

Sammy Maystone
smaystone4@fake.com | N VO | C E

# invoice_number_1

 

Date: Jan 1, 2022
Bill To: Ship To: P tT Net 30d
aymen erms: le ays
Taylor Riddel Taylor's Store y y
triddel9@fake.com triddel9@fake.com Due Date: Jan 31, 2022
PO Number: po_number_123
Balance Due: $127.50
cia Rate Amount
Service A 12 $10.00 $120.00
Description: Repair
Notes: Replaced capacitor
Parts: 1 x cap_a
Tax: 0.2%
Service B 5 $1.50 $7.50
Description: Cleaning
Notes: Removed debris from case
Parts: 2 x shop supplies
Tax: 0.4%
Subtotal: $127.50
Tax (0%): $0.00
Total: $127.50
Notes:

All names are fictitious and in no way represent real people or places.



```

</details>

**after**  sammymaystone.png as input (directly passed to tesseract)

<details>


```
Sammy Maystone                                                                                   | Ny           |      E
smaystone4@fake.com

                                                                                                 # invoice_number_1



                                                                                         Date:               Jan 1, 2022
Bill To:                      Ship To:
           .                                                                 Payment Terms:                  Net 30 days
Taylor Riddel                 Taylor's Store
triddel9@fake.com             triddel9@fake.com                                     Due Date:               Jan 31, 2022

                                                                                   PO Number:             po_number_123

                                                                             Balance Due:                      $127.50




Item                                                                    Quantity                   Rate          Amount

Service A                                                               12                       $10.00         $120.00
Description: Repair
Notes: Replaced capacitor
Parts: 1 x cap_a
Tax: 0.2%

Service B                                                               5                         $1.50             $7.50
Description: Cleaning
Notes: Removed debris from case
Parts: 2 x shop supplies
Tax: 0.4%




                                                                                     Subtotal:                  $127.50

                                                                                     Tax (0%):                      $0.00

                                                                                        Total:                  $127.50




Notes:
All namesarefictitious and in no way represent real people or places.




```

</details>

**after**  sammymaystone.pdf


<details>


```
Sammy Maystone                                                                                                   |      F
smaystone4@fake.com                                                                               | N Vy

                                                                                                  # invoice_number_1



                                                                                          Date:               Jan 1, 2022
Bill To:                      Ship To:
        ,                                                                     Payment Terms:                  Net 30 days
Taylor Riddel                 Taylor's Store
triddel9@fake.com             triddel9@fake.com                                      Due Date:               Jan 31, 2022

                                                                                    PO Number:             po_number_123

                                                                              Balance Due:                      $127.50




LCa                                                                      Quantity                   Rate          Amount

Service A                                                                12                       $10.00         $120.00
Description: Repair
Notes: Replaced capacitor
Parts: 1 x cap_a
Tax: 0.2%

Service B                                                                5                         $1.50             $7.50
Description: Cleaning
Notes: Removeddebris from case
Parts: 2 x shop supplies
Tax: 0.4%




                                                                                      Subtotal:                  $127.50

                                                                                      Tax (0%):                      $0.00

                                                                                         Total:                  $127.50




Notes:
All namesare fictitious and in no way represent real people or places.



```

</details>

**Comment by bosd** [comment]
## FlipkartInvoice

**before** flipkartinvoice.png as input


<details>


```

 

Sold By : WS Retail Services Pvt L

td.,
KACHERAKANILALL VILAGE. JADHCENAUALL HOBLL HOSKOTE TALL ani, Kerma In S067

 

 

 

 

 

 

 

‘Order ID: 00304175096047380001 Billing Address ‘Shipping Address
‘Anushrut Singh “Anushrut Singh
ores Datee as sa 3/64, Vishwas Khand,Gomt Nagar, 364, Vishwas Khand.Gomt Nagar, “apis ince amd
pene infor
. enclave. enclave. ane
VAT: Lseknow 226010 Uttar Pradesh _Licknow 226010 Uttar Pradesh
Service tax #: AAACWE725FSD001 Phone: 6756390642 Phone: 8756390642,
Product Tite ty Price®) Taxi) Tax(®) Total®)
Memory cars ‘SanDisk Ultra 16 GB MicroSDHC Class + 2761 1450% 40.39 31900
FN: 10 48 MBs Memory Card cst
ACCEASESGPEQNEPS Presi Promina 108
woo: vM72616
Total 1 278.61 40.39 319.00
Grand Total = 319.00
ise sed i ‘WS Reta Series Pvt id

 



```

</details>

**after** flipkartinvoice.png passed cleanly

<details>


```
                                                                                                           [invoice No 1 WBRWELD20151000982550}
Sold By : WSRetail Services Pvt. Ltd.,
Warehouse Address: NO. 421 & 43, KACHERAKANAHALLI VILLAGE,JADIGENAHALLIHOBLI, HOSKOTE TALUK, Bangalore. Karnataka, India $6067
 Order ID: 0D304175096047380001 Billing Address                                      Shipping Address
                i                              ‘Anushrut Singh                       ‘Anushrut Singh
 Order Date: 15-10-2015                        3/64, Vishwas Khand,Gomti Nagar,, 3/64, Vishwas Khand,Gomti Nagar,,                    “Keep this invoice and
 Invoice Date: 20-10-2015                      nearFun republic mall and nehu         near Funrepublic mall and nehru                  ‘mamebor
           .                                   enclave.                              enclave.                                           rantppt
 VATITIN: 29671                                Lucknow 226010 Uttar Pradesh           Lucknow 226010 Uttar Pradesh
 Service tax #: AAACW8725FSD001 Phone: 8756390642                                     Phone: 8756390642
 Product                          Title                                                    Qty Price(®) —Tax(%)                  —Tax(@)_—_      Total (®)
 Memory cards                     ‘SanDisk Ultra 16 GB MicroSDHC Class                        4         27861 1450%                 4039           319.00
 FSN:                             10 48 MB/s Memory Card                                                                   CsT
 ACCEASESGPBOJ9P9 Priceix inclusivofPromotionalDiscountofR -40.00
 WID: VM72516
                                                                                 Total 4 278.61                                  40.39           319.00

                                                                                                       Grand Total                       = 319.00
odRa vce    PELE
          k
         Ne

                                                                                                                                    (Authorized Signatory)




```

</details>

**after** flipkartinvoice.pdf 
### todo end of commute
<details>


```


```

</details>

**Comment by bosd** [comment]
## Amazonwebservices

**Before** amazonwebservices.png

<details>


```

Amazon bared Services lwolce:

 

invoice Summary —
os =r
— ane
carnage: TOTAL AMOUNT DUE ON August 3, 2016 sunt
caesawchony
es A ooo,
This invoice is for the billing period July 1 - July 31 , 2014
See nett neater eee
‘Summary
{AWS Serve Charges, sat
cna aa
=
= Sj

‘AWS Data Transfer
‘charges

‘Amazon Elastic Compute Cloud
‘charges
var

‘Amazon Glacier
‘charges

‘Amazon Simple Storage Service
‘charges

estimated US eles tx VAT GST

“May incase
Gch OT torent atc then ot nce


```

</details>

**After** passed image cleanly:


<details>


```
 m am azo  D) n                                       Amazon Web Services Invoice
mA webservices“                                       Era orate buteurANSaccross coca
fecoun umber                                          Invoice Summary
296664039561                                          ToceNumber:                                                         az1ea017
                                                      bicDate:                                                        August, 2014
Alto Address:
ATTN: MiveLabs Limited                                TOTAL AMOUNT DUEON Aux gust 3,2014
                                                                                   3,                                        $4.11
93B Sai Yu Chung
Yuen Long,N.T., 0000, HK

This invoice is forthebilling period July 1 - July 31 , 2014
Creating rm Amazon Web Services,we're wingtoprovideyouwäh anelectronic invoice foryour use ofAWS services. Adational intormation
‘egarcng yourpil, dividual service chargedetaia, and youracounthistoryare avadabieon theAccountAcivity Page.
 Summary
   AWSService Charges                                                                                                         $4.11
     charges                                                                                                                   sant
     recs                                                                                                                      ‘sooo
     Text                                                                                                                      sooo


 Detail
   AWSData Transfer                                                                                                           $0.01
     Charges                                                                                                                   ‘soot
     var"                                                                                                                      ‘sooo
   ‘AmazonElastic Compute Cloud                                                                                               $1.87
     (Charges                                                                                                                  S187
     var"                                                                                                                      sooo
   Amazon Glacier                                                                                                             $2.22
     Charges                                                                                                                   s22
     var"                                                                                                                      sooo
   ‘AmazonSimple Storage Service                                                                                              $0.01
     (Charges                                                                                                                  ‘soot
     var"                                                                                                                      sooo

Mayinclude
Tiss        estimated USe salestax, VAT, GST and CT
        notaVATinvoic                                                                           Servicoprovier:
                                                                                                {ottobe  ueserpaymentemtance)
~ Checkthe GSTstatement attached atthe endofthis Invoice                                        reenSarees be
fordette                                                                                        ‘StateWn 38108-5210, US



```

</details>

**After** passing the pdf as an input.
Thus using the pre-processing from the tesseract wrapper.

<details>


```
             asap
iis
 Cio amazon
      b  oe                                                  Amazon WebServices Invoice
                we    services                               Email or talk to us about your AWS accountorbill, visit aws.amazon.com/contact-us/
Account number:                                                    z

296664039561                                                 Invoice Summary
                                                             Invoice Number:                                                                42183017
                                                             Invoice Date:                                                            August 3 , 2014
Bill to Address:
ATTN: iViveLabs Limited                                      TOTAL AMOUNTDUE ON August3 , 2014                                                 $4.11
93B Sai Yu Chung
Yuen Long, N.T., 0000, HK




This invoice is for the billing period July 1 - July 31 , 2014
Greetings from Amazon Web Services, we're writing to provide you with an electronic invoice for your use of AWS services. Additional information
regarding your bill, individual service charge details, and your account history are available on the Account Activity Page.


 Summary
    AWS Service Charges                                                                                                                        $4.11
       Charges                                                                                                                                    $4.11
       Credits                                                                                                                                    $0.00
       Tax *                                                                                                                                      $0.00

 Total for this invoice                                                                                                                        $4.11


 Detail
    AWS Data Transfer                                                                                                                          $0.01
       Charges                                                                                                                                    $0.01
       VAT **                                                                                                                                     $0.00

    Amazon Elastic Compute Cloud                                                                                                               $1.87
       Charges                                                                                                                                    $1.87
       VAT **                                                                                                                                     $0.00

    Amazon Glacier                                                                                                                             $2.22
       Charges                                                                                                                                    $2.22
       VAT **                                                                                                                                     $0.00

    Amazon Simple Storage Service                                                                                                              $0.01
       Charges                                                                                                                                    $0.01
       VAT **                                                                                                                                     $0.00




* Mayinclude estimated US sales tax, VAT, GST and CT                                                        Service Provider:
** This is not a VAT invoice                                                                                (Not to be used for paymentremittance)
*** Check the GST statement attached at the end ofthis Invoice                                              Amazon WebServices,Inc.
                                                                                                            410 Terry Ave North
for details
                                                                                                            Seattle, WA 98109-5210, US




                                                                                                                                                          1



```

</details>

**Comment by bosd** [comment]
@rmilecki Can we merge this one?

File: README.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ A command line tool and Python library to support your accounting
 process.
 
 1. extracts text from PDF files using different techniques, like
-   `pdftotext`, `text`, `pdfminer` or OCR -- `tesseract`, `tesseract4` or
+   `pdftotext`, `pdfminer` or OCR -- `tesseract`, or
```
Comment (rmilecki) [inline README.md:11]: Please leave `text` here :)

File: src/invoice2data/input/tesseract.py [hunk 2]
```diff
@@ -1,29 +1,42 @@
...
+from subprocess import Popen, PIPE, STDOUT, CalledProcessError, TimeoutExpired
+from subprocess import run
+from pathlib import Path
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def to_text(path, language="en"):
```
Comment (rmilecki) [inline src/invoice2data/input/tesseract.py:15]: This parameter seems to be ignored as below you do:
```
language = get_languages()
```
Can it be just dropped?

File: src/invoice2data/main.py [hunk 3]
```diff
@@ -26,7 +25,6 @@
 input_mapping = {
     "pdftotext": pdftotext,
     "tesseract": tesseract,
-    "tesseract4": tesseract4,
```
Comment (rmilecki) [inline src/invoice2data/main.py:29]: Does it make sense to keep `tesseract4` mapped to `tesseract` just for some kind of backward compatibility? I'm not sure if it's worth it, just wondering.

File: README.md (hunk 1 above)
Comment (bosd) [reply README.md:11]: Fine by me ;) Will fix this merge conflict

File: src/invoice2data/input/tesseract.py (hunk 2 above)
Comment (bosd) [reply src/invoice2data/input/tesseract.py:15]: I left it in for "safety". but if the get_language function is failling, propably tesseract will fail anyway. 
Removing this, as keeping it in could cause confusion.

File: src/invoice2data/main.py (hunk 3 above)
Comment (bosd) [reply src/invoice2data/main.py:32]: In my opinion it's not worth it. was'nt well documented and functional anyway.
There's no active implementation with the old teseract known to me.
We can always add it if we got some feedback from someone who needs it.

**Comment by rmilecki** [comment]
Please clean up those commits (there are 6 commits right now - 2 merges in it) and merge! :)

**Review by rmilecki — APPROVED** [review]

**Pull request merged**